File: src/agent_extensions/plugins/noesis/scripts/check_conversation_id.py
# /// script
# dependencies = [
#     "pydantic",
# ]
# ///

"""Check conversation_id presence in transcript and knowledge graph."""

import json
import sys
import uuid
from pathlib import Path

from models_transcript import CONVERSATION_ID_PATTERN


def extract_conversation_id(transcript_path: Path) -> str | None:
    with open(transcript_path) as f:
        first_line = f.readline()
    match = CONVERSATION_ID_PATTERN.match(first_line)
    return match.group(1) if match else None


def is_id_in_knowledge_graph(conversation_id: str, knowledge_graph_path: Path) -> bool:
    if not knowledge_graph_path.exists():
        return False
    text = knowledge_graph_path.read_text()
    return f'"conversation_id": "{conversation_id}"' in text


def prepend_conversation_id(conversation_id: str, transcript_path: Path) -> None:
    content = transcript_path.read_text()
    transcript_path.write_text(f"<!-- conversation_id: {conversation_id} -->\n{content}")


def main() -> None:
    if len(sys.argv) != 3:
        print(json.dumps({"status": "Error", "message": "Expected 2 arguments: <transcript_path> <knowledge_graph_path>"}))
        sys.exit(1)

    transcript_path = Path(sys.argv[1])
    knowledge_graph_path = Path(sys.argv[2])

    conversation_id = extract_conversation_id(transcript_path)

    if conversation_id is None:
        conversation_id = str(uuid.uuid4())
        prepend_conversation_id(conversation_id, transcript_path)
        result = {"status": "IdGenerated", "conversation_id": conversation_id}
    elif is_id_in_knowledge_graph(conversation_id, knowledge_graph_path):
        result = {"status": "ConversationAlreadyAdded", "conversation_id": conversation_id}
    else:
        result = {"status": "Ok", "conversation_id": conversation_id}

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: src/agent_extensions/plugins/noesis/scripts/transcript/check-conversation-id.ts
import { readFileSync, writeFileSync } from "fs";
import { existsSync } from "fs";
import { randomUUID } from "crypto";
import { parseArgs, requireFile, outputResult } from "../io.js";
import { CONVERSATION_ID_PATTERN } from "./types.js";

export function extractConversationId(transcriptPath: string): string | null {
  const content = readFileSync(transcriptPath, "utf-8");
  const firstLine = content.split("\n", 1)[0];
  const match = CONVERSATION_ID_PATTERN.exec(firstLine);
  return match ? match[1] : null;
}

export function isIdInKnowledgeGraph(
  conversationId: string,
  knowledgeGraphPath: string,
): boolean {
  if (!existsSync(knowledgeGraphPath)) {
    return false;
  }
  const text = readFileSync(knowledgeGraphPath, "utf-8");
  return text.includes(`"conversation_id": "${conversationId}"`);
}

export function prependConversationId(
  conversationId: string,
  transcriptPath: string,
): void {
  const content = readFileSync(transcriptPath, "utf-8");
  writeFileSync(
    transcriptPath,
    `<!-- conversation_id: ${conversationId} -->\n${content}`,
    "utf-8",
  );
}

function main(): void {
  const args = parseArgs(["transcript_path", "knowledge_graph_path"]);
  requireFile(args["transcript_path"]);

  const conversationId = extractConversationId(args["transcript_path"]);

  if (conversationId === null) {
    const newId = randomUUID();
    prependConversationId(newId, args["transcript_path"]);
    outputResult({ status: "IdGenerated", conversation_id: newId });
  } else if (
    isIdInKnowledgeGraph(conversationId, args["knowledge_graph_path"])
  ) {
    outputResult({
      status: "ConversationAlreadyAdded",
      conversation_id: conversationId,
    });
  } else {
    outputResult({ status: "Ok", conversation_id: conversationId });
  }
}

if (import.meta.main) {
  main();
}
